### PR TITLE
Adds Teleprods

### PR DESCRIPTION
**Pull request by PKPenguin321**
Crafted by sticking a bluespace crystal (artificial or real) onto a power cell-less stun prod, or by tablecrafting with the ingredients for a stunprod + a bluespace crystal (again, it can be artificial).

They are functionally similar to stunprods, requiring a power cell and having a limited number of charges.

When somebody is stunned by it, they are also teleported away to a random nearby area. This _can_ space them/put them into walls. The teleport works even if the stun doesn't (IE, if you hit a xeno it would teleport but they would not get stunned).

teleprod sprites

:cl: PKPenguin321
rscadd: Teleprods can now be constructed. Make them via tablecrafting, or by putting a bluespace crystal onto a stunprod. They warp the target to a random nearby area (which may include space/the inside of a wall) in addition to giving the victim a brief stun. Like stunprods, they require a power cell to function.
rscadd: Teleprods can have their crystal removed and be made back into stunprods by using them in your hand when they have no cell installed.
/:cl:

### CODE REVIEW

**Comment by PKPenguin321** [comment]
aaaand i forgot to make them use the new inhands

File: tgstation.dme [hunk 1]
```diff
@@ -922,7 +923,6 @@
 #include "code\modules\awaymissions\mission_code\spacebattle.dm"
 #include "code\modules\awaymissions\mission_code\stationCollision.dm"
 #include "code\modules\awaymissions\mission_code\wildwest.dm"
-#include "code\modules\client\asset_cache.dm"
```
Comment (optimumtact) [inline tgstation.dme]: h-h-hey what are you d-d-doing

**Comment by Zilenan91** [comment]
Can you remove the stun on it?

**Comment by PKPenguin321** [comment]
>Can you remove the stun on it?

i mean i could but why would i

File: tgstation.dme (hunk 1 above)
Comment (PKPenguin321) [inline tgstation.dme]: shhhh

Comment (JJRcop) [inline tgstation.dme]: This happens a lot, it's very likely he didn't mean to do it.

**Comment by Zilenan91** [comment]
Because it doesn't need or make sense to have a stun on it. Bluespace crystals don't really stun people, and the teleportation combined with the stun can fuck people over more than just being prodded does

**Comment by PKPenguin321** [comment]
it's a weapon
it fucking people over is kind of the idea ¯\\\_(ツ)_/¯

**Comment by Zilenan91** [comment]
Then make a stunprod if you want to stun people. The stun combined with the teleportation is too much.

**Comment by MMMiracles** [comment]
zilenan stop hating fun.

people are totally gonna be raiding the station teleporter more often now solely for those sweet, sweet crystals though.

**Comment by Incoming5643** [comment]
Stick with the prod

**Comment by GunHog** [comment]
Do not be lazy! Allow this to be done with proper Security batons as well.

**Comment by KorPhaeron** [comment]
Sounds pretty useless but fun

**Comment by optimumtact** [comment]
>gunhog wants a subtle security buff
COLOUR ME SUPRISED

**Comment by PKPenguin321** [comment]
>Do not be lazy! Allow this to be done with proper Security batons as well.

this would basically double the size of this pr, so nah
plus how would you even mount a bluespace crystal to a baton i mean cmon

**Comment by GunHog** [comment]
I do not play Security, Oranges - stop confusing me with Steelpoint. It did not make sense to me for the child item to have it, but not the parent. 

You would mount the crystal the same way you did for your piece of rod, wirecutters, and cable - by stuffing it between the terminals. If your super delicate crystal held in place with a sharp, metallic clamp which itself is tied to a metal rod with some uninsulated wire makes sense, then so would that. It does not really matter to me as a player, as I can much more easily access a stun prod than I could a baton.

**Comment by PKPenguin321** [comment]
i guess but again that would require a shitload more code and prods work fine anyways

**Comment by WJohn** [comment]
If it DOES stun them, I think it should do the stunning *after* it teleports them, that way you don't have them just drop the items they were holding and be too far to get them back. It'd be an intensely powerful inhand-stealing tool.

**Comment by KorPhaeron** [comment]
It requires a rare mining material+you give up the chance to kill reliably it's fine as is

**Comment by Zilenan91** [comment]
Or just decon one of the teleporters like the one on the AI sat.

**Comment by PKPenguin321** [comment]
balance police please go

**Comment by GunHog** [comment]
Oh god, the filthy greytide is going to breaking into the teleporter room each round. I did not even think about this.

**Comment by PKPenguin321** [comment]
maybe for like three rounds before the novelty wears off

**Comment by GunHog** [comment]
When I play RD, I should consider building a turret or two in there to discourage theft...

**Comment by Zilenan91** [comment]
do it and set it to laser. Cut the AI camera too so it doesn't deactivate it.

**Comment by Lumipharon** [comment]
So you'll be able to teleport yourself with this I assume (even if it stuns you if you're not hulk).
This seems fairly strong, and basically just incentivises ripping machines up to make weapons.

**Comment by GunHog** [comment]
Using the weapon on oneself makes it one-use for most except hulks and anti-drop implant users, as the stun happens first, causing the user to drop the weapon.

Still, it is a valid concern, even if the user takes the risk of stunning himself and losing the weapon first.

**Comment by PKPenguin321** [comment]
if anybody has a better idea for implementing these that doesnt require bluespace crystals, hit me with it either here or on irc
for now im just gonna leave it like this since i think you guys are really *really* **really** overestimating the balance aspects of this

**Comment by Dorsisdwarf** [comment]
This looks like everyone's overthought it so much

If you catch an assistant with one of these as sec, insta-perma them for tresspass, theft, and posession of a weapon. There's basically no way to legitimately have one of these. boom.

**Comment by Lumipharon** [comment]
printing a crystal with R&D, making one at xeno, or finding one at mining are all genuine ways to get bluespace crystals. Also looting off station teleporters.

**Comment by Stenthus** [comment]
how far away does it teleport people?

**Comment by Dorsisdwarf** [comment]
:s HOS ask the RD if they're handing out telecrystals to greyshirts

**Comment by dumpdavidson** [comment]
>Then make a stunprod if you want to stun people. The stun combined with the teleportation is too much.

But teleporting someone you just stunned away is worse than just stunning them. 

I agree with Kor here: useless but fun.

**Comment by Alek2ander** [comment]
You know you can throw the crystals, right?

**Comment by CosmicScientist** [comment]
@Alek2ander that uses up the crystal, from the looks of this you get infinite uses for tapping people with a rod, the only drawback is replacing/recharging the cell every-so-often (I assume as much as the regular stunprod).

**Comment by optimumtact** [comment]
we can always add crystal degradation if it proves to be too annoying.

**Comment by PKPenguin321** [comment]
@CosmicScientist Yep
@Stenthus It should warp between 1 and 15 tiles away. If a clown accidentally hits themselves with it out of clumsiness, they can warp up to 50 tiles away (honk)

**Comment by PKPenguin321** [comment]
Do not merge for now, mining changes got merged and changed the pathing of bluespace crystals, meaning this won't work until I update it
Also I'm gonna try to add the ability to remove the crystal

**Comment by duncathan** [comment]
I restarted your Travis build just for you @PKPenguin321

**Comment by PKPenguin321** [comment]
So yeah, conflicts fixed, also you can now remove the crystal from a teleprod and make it back into a stunprod by using it in your hand when it has no cell installed. @GunHog I also found a way to make batons work, too, but I'm gonna save it for a separate PR to avoid the monster of controversy that is additions security.

**Comment by Bawhoppen** [comment]
would be top-tier to have one that teleports every item on your character to a different random location


also should be able to teleport other crap you hit like chairs and shit

**Comment by PKPenguin321** [comment]
>would be top-tier to have one that teleports every item on your character to a different random location

would be hilarious, but not happening for both code and balance reasons

>also should be able to teleport other crap you hit like chairs and shit

nah, for the same reasons (plus "Who stole the particle accelerator?!" "Issue Report: Singularity can be warped away with a teleprod")

**Comment by KorPhaeron** [comment]
@PKPenguin321 

Could just make it check for anchored, which I'm hoping it already does for humans

**Comment by PKPenguin321** [comment]
Yeah, they do check for anchored on humans

This is also pretty much ready for review/merge

File: code/game/objects/items/weapons/teleprod.dm [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	if(status && user.disabilities & CLUMSY && prob(50))
+		user.visible_message("<span class='danger'>[user] accidentally hits themself with [src]!</span>", \
+							"<span class='userdanger'>You accidentally hit yourself with [src]!</span>")
+		user.Weaken(stunforce*3)
+		deductcharge(hitcost)
+		do_teleport(user, get_turf(user), 50)//honk honk
+		return
+	else
+		if(status)
+			if(!istype(M) && M.anchored)
```
Comment (duncathan) [inline code/game/objects/items/weapons/teleprod.dm]: Are you sure this should be `&&`?

Comment (PKPenguin321) [inline code/game/objects/items/weapons/teleprod.dm]: honestly i have no clue, i was hoping reviewers could tell me
i can do ingame tests real fast, hold on

Comment (GunHog) [inline code/game/objects/items/weapons/teleprod.dm]: Testing is fun :)

Comment (PKPenguin321) [inline code/game/objects/items/weapons/teleprod.dm]: okay so after fixing teleprods and testing again, i can confirm the anchor check works. if a mob is buckled (anchored) and hit with a teleprod, they automatically unbuckle, meaning there's no problemo.

the only real issue i found was that if you're clumsy and accidentally hit yourself when trying to hit someone else, they still teleport even though they're not stunned. not sure how to fix this one.

**Comment by PKPenguin321** [comment]
WOW @Cheridan 
RIGHT AS I MAKE MY COMMIT TO FIX THE BUGS YOU MERGE MY OTHER PR WHICH MAKES THE INHANDS CONFLICT
REEEEEEEEEEEEEEEEEEEEEEEEEEEEEE

**Comment by PKPenguin321** [comment]
fucking updating the inhands was harder than the actual coding, fucking git :rage:

**Pull request merged**